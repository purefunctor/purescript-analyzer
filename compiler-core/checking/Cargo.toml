[package]
name = "checking"
version = "0.1.0"
edition = "2024"

[dependencies]
files = { version = "0.1.0", path = "../files" }
indexing = { version = "0.1.0", path = "../indexing" }
interner = { version = "0.1.0", path = "../interner" }
itertools = "0.14.0"
la-arena = "0.3.1"
lexing = { version = "0.1.0", path = "../lexing" }
lowering = { version = "0.1.0", path = "../lowering" }
parsing = { version = "0.1.0", path = "../parsing" }
rowan = "0.16.1"
smol_str = "0.3.2"
syntax = { version = "0.1.0", path = "../syntax" }
