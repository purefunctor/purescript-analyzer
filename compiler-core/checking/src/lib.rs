pub mod check;
pub mod core;
pub mod debruijn;
