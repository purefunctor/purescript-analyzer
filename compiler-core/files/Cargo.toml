[package]
name = "files"
version = "0.1.0"
edition = "2024"

[dependencies]
indexmap = "2.7.1"
la-arena = "0.3.1"
rustc-hash = "2.1.1"
