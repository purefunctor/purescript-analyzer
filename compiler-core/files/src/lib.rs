use core::str;
use std::sync::Arc;

use indexmap::IndexMap;
use la_arena::{Idx, RawIdx};
use rustc_hash::FxBuildHasher;

#[derive(Debug, Default)]
pub struct Files {
    files: IndexMap<Arc<str>, Arc<str>, FxBuildHasher>,
}

pub struct File;

pub type FileId = Idx<File>;

impl Files {
    pub fn insert(&mut self, k: impl Into<Arc<str>>, v: impl Into<Arc<str>>) -> FileId {
        let k = k.into();
        let v = v.into();
        let (index, _) = self.files.insert_full(k, v);
        Idx::from_raw(RawIdx::from_u32(index as u32))
    }

    pub fn id(&self, k: &str) -> Option<FileId> {
        self.files.get_full(k).map(|(index, _, _)| Idx::from_raw(RawIdx::from_u32(index as u32)))
    }

    pub fn path(&self, id: FileId) -> Arc<str> {
        let index = id.into_raw().into_u32() as usize;
        let (path, _) =
            self.files.get_index(index).expect("invariant violated: expected valid FileId");
        Arc::clone(path)
    }

    pub fn content(&self, id: FileId) -> Arc<str> {
        let index = id.into_raw().into_u32() as usize;
        let (_, contents) =
            self.files.get_index(index).expect("invariant violated: expected valid FileId");
        Arc::clone(contents)
    }
}

#[cfg(test)]
mod tests {
    use super::Files;

    #[test]
    fn test_basic() {
        let mut files = Files::default();

        let k = "src/Main.purs";
        let v = "module Main where\n\n";

        let id = files.insert(k, v);

        assert_eq!(files.id(k), Some(id));
        assert_eq!(files.path(id).as_ref(), k);
        assert_eq!(files.content(id).as_ref(), v);
    }
}
