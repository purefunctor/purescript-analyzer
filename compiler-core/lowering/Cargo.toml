[package]
name = "lowering"
version = "0.1.0"
edition = "2021"

[dependencies]
hashbrown = "0.15.2"
indexing = { version = "0.1.0", path = "../indexing" }
indexmap = "2.7.1"
itertools = "0.14.0"
la-arena = "0.3.1"
paste = "1.0.15"
rowan = "0.16.1"
rustc-hash = "2.1.1"
smol_str = "0.3.2"
syntax = { version = "0.1.0", path = "../syntax" }

[dev-dependencies]
insta = "1.42.1"
lexing = { path = "../lexing" }
parsing = { path = "../parsing" }
test_each_file = "0.3.4"
