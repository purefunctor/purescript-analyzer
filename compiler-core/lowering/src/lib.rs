mod algorithm;

pub mod intermediate;
pub mod scope;
pub mod source;

pub use intermediate::*;
pub use scope::*;
pub use source::*;

use indexing::FullIndexedModule;
use syntax::cst;

#[derive(Debug, PartialEq, Eq)]
pub struct FullLoweredModule {
    pub intermediate: Intermediate,
    pub source: LoweringSource,
    pub graph: LoweringGraph,
    pub graph_info: LoweringGraphInfo,
}

pub fn lower_module(module: &cst::Module, indexed: &FullIndexedModule) -> FullLoweredModule {
    let algorithm::State { intermediate, source, graph, graph_info, .. } =
        algorithm::lower_module(module, indexed);
    FullLoweredModule { intermediate, source, graph, graph_info }
}
