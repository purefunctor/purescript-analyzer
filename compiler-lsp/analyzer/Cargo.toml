[package]
name = "analyzer"
version = "0.1.0"
edition = "2024"

[dependencies]
async-lsp = { version = "0.2.2", features = ["tokio"] }
building = { version = "0.1.0", path = "../../compiler-core/building" }
files = { version = "0.1.0", path = "../../compiler-core/files" }
indexing = { version = "0.1.0", path = "../../compiler-core/indexing" }
itertools = "0.14.0"
la-arena = "0.3.1"
lexing = { version = "0.1.0", path = "../../compiler-core/lexing" }
line-index = "0.1.2"
lowering = { version = "0.1.0", path = "../../compiler-core/lowering" }
parsing = { version = "0.1.0", path = "../../compiler-core/parsing" }
resolving = { version = "0.1.0", path = "../../compiler-core/resolving" }
rowan = "0.16.1"
serde = { version = "1.0.219", features = ["derive"] }
serde_json = "1.0.140"
smol_str = "0.3.2"
strsim = "0.11.1"
syntax = { version = "0.1.0", path = "../../compiler-core/syntax" }
tracing = "0.1.41"
