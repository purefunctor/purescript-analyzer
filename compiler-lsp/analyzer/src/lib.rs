pub mod completion;
pub mod definition;
pub mod extract;
pub mod hover;
pub mod locate;
pub mod references;

pub use building::{QueryEngine, prim};
pub use files::Files;
