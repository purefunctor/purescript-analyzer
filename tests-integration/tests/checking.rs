use std::{fmt::Write, num::NonZeroU32};

use analyzer::{QueryEngine, prim};
use checking::{
    check::{
        CheckContext, CheckState, quantify,
        unification::{self, UnificationState},
    },
    core::{Type, TypeId, Variable, debruijn, pretty},
};
use files::{FileId, Files};
use lowering::TypeVariableBindingId;

struct ContextState<'r> {
    context: CheckContext<'r, QueryEngine>,
    state: CheckState,
}

impl<'a> ContextState<'a> {
    fn new(engine: &'a QueryEngine, id: FileId) -> ContextState<'a> {
        let mut state = CheckState::default();
        let context = CheckContext::new(engine, &mut state, id).unwrap();
        ContextState { state, context }
    }
}

trait CheckStateExt {
    fn bound_variable(&mut self, index: u32) -> TypeId;

    fn function(&mut self, argument: TypeId, result: TypeId) -> TypeId;
}

impl CheckStateExt for CheckState {
    fn bound_variable(&mut self, index: u32) -> TypeId {
        let var = Variable::Bound(debruijn::Index(index));
        self.storage.intern(Type::Variable(var))
    }

    fn function(&mut self, argument: TypeId, result: TypeId) -> TypeId {
        self.storage.intern(Type::Function(argument, result))
    }
}

fn empty_engine() -> (QueryEngine, FileId) {
    let mut engine = QueryEngine::default();
    let mut files = Files::default();
    prim::configure(&mut engine, &mut files);

    let id = files.insert("Main.purs", "module Main where\n\n");
    let content = files.content(id);
    engine.set_content(id, content);

    (engine, id)
}

const FAKE_NONZERO_1: NonZeroU32 = NonZeroU32::new(1).unwrap();
const FAKE_NONZERO_2: NonZeroU32 = NonZeroU32::new(2).unwrap();

#[test]
fn test_solve_simple() {
    let (engine, id) = empty_engine();
    let ContextState { ref context, ref mut state } = ContextState::new(&engine, id);

    // [a :: Int, b :: String]
    state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_1), context.prim.int);
    state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_2), context.prim.string);

    let unification = state.fresh_unification_type(context);
    let Type::Unification(unification_id) = state.storage[unification] else {
        unreachable!("invariant violated");
    };

    unification::solve(state, context, unification_id, context.prim.symbol).unwrap();

    let entry = *state.unification.get(unification_id);
    let UnificationState::Solved(solution) = entry.state else {
        unreachable!("invariant violated");
    };

    let solution = pretty::print_local(state, context, solution);
    let kind = pretty::print_local(state, context, entry.kind);

    let snapshot = format!("{solution} :: {kind}");
    insta::assert_snapshot!(snapshot);
}

#[test]
fn test_solve_bound() {
    let (engine, id) = empty_engine();
    let ContextState { ref context, ref mut state } = ContextState::new(&engine, id);

    // [a :: Int, b :: String]
    state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_1), context.prim.int);
    state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_2), context.prim.string);

    let unification = state.fresh_unification_type(context);
    let Type::Unification(unification_id) = state.storage[unification] else {
        unreachable!("invariant violated");
    };

    let bound_b = state.bound_variable(0);
    let bound_a = state.bound_variable(1);
    let b_to_a = state.function(bound_b, bound_a);

    unification::solve(state, context, unification_id, b_to_a).unwrap();

    let entry = *state.unification.get(unification_id);
    let UnificationState::Solved(solution) = entry.state else {
        unreachable!("invariant violated");
    };

    let solution = pretty::print_local(state, context, solution);
    let kind = pretty::print_local(state, context, entry.kind);

    let snapshot = format!("{solution} :: {kind}");
    insta::assert_snapshot!(snapshot);
}

#[test]
fn test_solve_invalid() {
    let (engine, id) = empty_engine();
    let ContextState { ref context, ref mut state } = ContextState::new(&engine, id);

    // [a :: Int]
    state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_1), context.prim.int);

    let unification = state.fresh_unification_type(context);
    let Type::Unification(unification_id) = state.storage[unification] else {
        unreachable!("invariant violated");
    };

    // [a :: Int, b :: String]
    let level = state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_2), context.prim.string);

    let bound_b = state.bound_variable(0);
    let bound_a = state.bound_variable(1);
    let b_to_a = state.function(bound_b, bound_a);

    state.unbind(level);

    let solve_result = unification::solve(state, context, unification_id, b_to_a);
    assert!(solve_result.is_none());
}

#[test]
fn test_solve_promotion() {
    let (engine, id) = empty_engine();
    let ContextState { ref context, ref mut state } = ContextState::new(&engine, id);

    // [a :: Int]
    state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_1), context.prim.int);

    let unification_a = state.fresh_unification_type(context);
    let Type::Unification(unification_id) = state.storage[unification_a] else {
        unreachable!("invariant violated");
    };

    // [a :: Int, b :: String]
    state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_2), context.prim.string);

    let unification_a_b = state.fresh_unification_type(context);
    unification::solve(state, context, unification_id, unification_a_b).unwrap();

    let mut snapshot = String::default();

    let entries: Vec<_> = state.unification.iter().copied().collect();
    for (index, entry) in entries.iter().enumerate() {
        let UnificationState::Solved(solution) = entry.state else { continue };
        let domain = entry.domain;
        let solution = pretty::print_local(state, context, solution);
        writeln!(snapshot, "?{index}[{domain}] := {solution}").unwrap();
    }

    insta::assert_snapshot!(snapshot);
}

#[test]
fn test_quantify_simple() {
    let (engine, id) = empty_engine();
    let ContextState { ref context, ref mut state } = ContextState::new(&engine, id);

    let unification_a = state.fresh_unification_type(context);
    let unification_b = state.fresh_unification_type(context);

    let function = state.storage.intern(Type::Function(unification_a, unification_b));
    let quantified = quantify::quantify(state, function).unwrap();

    let mut snapshot = String::default();

    let quantified = pretty::print_local(state, context, quantified);
    writeln!(snapshot, "{quantified}").unwrap();

    insta::assert_snapshot!(snapshot)
}

#[test]
fn test_quantify_polykind() {
    let (engine, id) = empty_engine();
    let ContextState { ref context, ref mut state } = ContextState::new(&engine, id);

    let unification = state.fresh_unification(context);
    let quantified = quantify::quantify(state, unification).unwrap();

    let mut snapshot = String::default();

    let quantified = pretty::print_local(state, context, quantified);
    writeln!(snapshot, "{quantified}").unwrap();

    insta::assert_snapshot!(snapshot)
}

#[test]
fn test_quantify_ordering() {
    let (engine, id) = empty_engine();
    let ContextState { ref context, ref mut state } = ContextState::new(&engine, id);

    state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_1), context.prim.t);
    let unification_a = state.fresh_unification_type(context);

    state.bind_forall(TypeVariableBindingId::new(FAKE_NONZERO_2), context.prim.t);
    let unification_b = state.fresh_unification_type(context);

    let function = state.storage.intern(Type::Function(unification_b, unification_a));
    let quantified = quantify::quantify(state, function).unwrap();

    let mut snapshot = String::default();

    let quantified = pretty::print_local(state, context, quantified);
    writeln!(snapshot, "{quantified}").unwrap();

    insta::assert_snapshot!(snapshot)
}

#[test]
fn test_quantify_scoped() {
    let (engine, id) = empty_engine();
    let ContextState { ref context, ref mut state } = ContextState::new(&engine, id);

    let unification_0 = state.fresh_unification_type(context);
    let unification_1 = state.fresh_unification_kinded(unification_0);
    let unification_2 = state.fresh_unification_kinded(unification_1);

    let quantified = quantify::quantify(state, unification_2).unwrap();

    let mut snapshot = String::default();

    let quantified = pretty::print_local(state, context, quantified);
    writeln!(snapshot, "{quantified}").unwrap();

    insta::assert_snapshot!(snapshot)
}

#[test]
fn test_quantify_multiple_scoped() {
    let (engine, id) = empty_engine();
    let ContextState { ref context, ref mut state } = ContextState::new(&engine, id);

    let unification_0 = state.fresh_unification_type(context);
    let unification_1 = state.fresh_unification_kinded(unification_0);
    let unification_2 = state.fresh_unification_kinded(unification_1);

    let unification_3 = state.fresh_unification_type(context);
    let unification_4 = state.fresh_unification_kinded(unification_3);
    let unification_5 = state.fresh_unification_kinded(unification_4);

    let function = state.storage.intern(Type::Function(unification_2, unification_5));
    let quantified = quantify::quantify(state, function).unwrap();

    let mut snapshot = String::default();

    let quantified = pretty::print_local(state, context, quantified);
    writeln!(snapshot, "{quantified}").unwrap();

    insta::assert_snapshot!(snapshot)
}

#[test]
fn test_manual() {
    let (engine, id) = empty_engine();

    engine.set_content(id, "module Main where\n\nforeign import data T :: Proxy 123");

    let resolved = engine.resolved(id).unwrap();
    let checked = engine.checked(id).unwrap();

    let (_, id) = resolved.locals.lookup_type("T").unwrap();
    let id = checked.lookup_type(id).unwrap();

    eprintln!("{}", pretty::print_global(&engine, id));
}
