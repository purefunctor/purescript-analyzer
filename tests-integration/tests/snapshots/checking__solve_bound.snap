---
source: tests-integration/tests/checking.rs
expression: snapshot
---
(*0 -> *1) :: Type
